sdk/lib/rtl/threadpool.c

```diff
@@ -0,0 +1,3485 @@
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+
+#ifdef __REACTOS__
+#include <rtl_vista.h>
+
+#include "wine/list.h"
+#include <debug.h>
+#ifndef ARRAY_SIZE
+#define ARRAY_SIZE(_x) (sizeof((_x))/sizeof((_x)[0]))
+#endif
+
+/* Winternal ?*/
+
+typedef struct _THREAD_NAME_INFORMATION
+{
+    UNICODE_STRING ThreadName;
+} THREAD_NAME_INFORMATION, *PTHREAD_NAME_INFORMATION;
+
+typedef void (CALLBACK *PNTAPCFUNC)(ULONG_PTR,ULONG_PTR,ULONG_PTR); /* FIXME: not the right name */
+typedef void (CALLBACK *PRTL_THREAD_START_ROUTINE)(LPVOID); /* FIXME: not the right name */
+typedef DWORD (CALLBACK *PRTL_WORK_ITEM_ROUTINE)(LPVOID); /* FIXME: not the right name */
+typedef void (NTAPI *RTL_WAITORTIMERCALLBACKFUNC)(PVOID,BOOLEAN); /* FIXME: not the right name */
+typedef VOID (CALLBACK *PRTL_OVERLAPPED_COMPLETION_ROUTINE)(DWORD,DWORD,LPVOID);
+
+typedef void (CALLBACK *PTP_IO_CALLBACK)(PTP_CALLBACK_INSTANCE,void*,void*,IO_STATUS_BLOCK*,PTP_IO);
+
+// More winternl maybe
+NTSYSAPI NTSTATUS  WINAPI TpSimpleTryPost(PTP_SIMPLE_CALLBACK,PVOID,TP_CALLBACK_ENVIRON *);
+
+// Redefines
+#define PRTL_WORK_ITEM_ROUTINE WORKERCALLBACKFUNC
+
+#define CRITICAL_SECTION RTL_CRITICAL_SECTION
+#define GetProcessHeap() RtlGetProcessHeap()
+#define GetCurrentProcess()   NtCurrentProcess()
+#define GetCurrentThread()    NtCurrentThread()
+#define GetCurrentThreadId()  HandleToULong(NtCurrentTeb()->ClientId.UniqueThread)
+#else
+#include <assert.h>
+#include <stdarg.h>
+#include <limits.h>
+
+#include "ntstatus.h"
+#define WIN32_NO_STATUS
+#include "winternl.h"
+
+#include "wine/debug.h"
+#include "wine/list.h"
+
+#include "ntdll_misc.h"
+#endif
+
+WINE_DEFAULT_DEBUG_CHANNEL(threadpool);
+
+/*
+ * Old thread pooling API
+ */
+
+struct rtl_work_item
+{
+    PRTL_WORK_ITEM_ROUTINE function;
+    PVOID context;
+};
+
+#define EXPIRE_NEVER       (~(ULONGLONG)0)
+#define TIMER_QUEUE_MAGIC  0x516d6954   /* TimQ */
+
+static RTL_CRITICAL_SECTION_DEBUG critsect_compl_debug;
+
+static struct
+{
+    HANDLE                  compl_port;
+    RTL_CRITICAL_SECTION    threadpool_compl_cs;
+}
+old_threadpool =
+{
+    NULL,                                       /* compl_port */
+    { &critsect_compl_debug, -1, 0, 0, 0, 0 },  /* threadpool_compl_cs */
```

Comment:
I know this is winesynced code, but application verifier is very unhappy with all these statically allocated critical sections.

Reply:
Alright i THINK i can do it, but i have very much underestimated how complicated of a change this is going to require.